File: node/tests/ffi-native-smoke.test.ts
/**
 * Real-FFI smoke tests for haiinpm.
 *
 * Two tests, one per backend, both loading the real haiinpm native addon
 * and exercising `saveMemory("...")` end-to-end:
 *
 * 1. **Remote** (`saveMemory round-trips through the real native binding`)
 *    — hosted production path. Sets `JACS_DEFAULT_STORAGE=remote` so the
 *    FFI builds a `RemoteJacsProvider`, signs locally, POSTs to a
 *    `node:http.createServer` mock, and reads the server-issued key from
 *    the response. Verifies the mock saw a `POST /api/v1/records` with
 *    signed markdown bytes (plaintext plus the JACS signature footer).
 *
 * 2. **Local** (`saveMemory persists locally without HTTP traffic`) — dev
 *    default path (the JACS binding writes `jacs_default_storage: "fs"`). Bootstraps
 *    a fresh agent, sets `JACS_DEFAULT_STORAGE=fs`, signs locally, writes to
 *    disk, and returns a client-side `{jacsId}:{jacsVersion}` key. Verifies
 *    the doc round-trips via `getRecordBytes(key)`.
 *
 * Together these cover the only two backends production and dev users
 * actually exercise.
 *
 * Skipped cleanly when:
 * - haiinpm is not built / installable (try-import + describe.skip).
 * - The JACS Node binding isn't available to bootstrap a test agent.
 *
 * Per PRD docs/haiai/JACS_DOCUMENT_STORE_FFI_PRD.md §5.5: real
 * `node:http.createServer` (no fetch-level mock). The traffic is Rust
 * `reqwest` running INSIDE the haiinpm native binding, which only a real
 * listening socket can intercept.
 */

import { afterEach, beforeEach, describe, expect, it, type TaskContext } from 'vitest';
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { existsSync, mkdtempSync, readFileSync, realpathSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { createRequire } from 'node:module';

const dynamicRequire = createRequire(import.meta.url);

let haiinpm: { HaiClient: new (config: string) => unknown } | null = null;
let loadError: unknown = null;
try {
  haiinpm = dynamicRequire('haiinpm') as { HaiClient: new (config: string) => unknown };
} catch (err) {
  loadError = err;
}

const describeWhenAvailable = haiinpm ? describe : describe.skip;

// `LocalJacsProvider::store_signed_text` returns the key as
// `{jacsId}:{jacsVersion}` where both halves are JACS UUIDs. This regex
// matches that exact shape so the local-path test asserts on the key
// *structure* (not a specific value, which would change every run).
const LOCAL_KEY_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

/**
 * Bootstrap or locate the JACS agent config. Two paths:
 * 1. Pre-baked agent dir via `JACS_SMOKE_AGENT_DIR` (preferred for CI —
 *    setup writes the agent once and all three smoke tests share it).
 * 2. In-process bootstrap via `@hai.ai/jacs` (local dev). Returns null
 *    when the JACS Node bindings aren't installed; caller should
 *    `ctx.skip()` to mark the test as SKIPPED (not PASSED with zero
 *    assertions, which a bare `return` produces in vitest).
 */
function resolveJacsAgentConfig(workdir: string): string | null {
  workdir = realpathSync(workdir);

  const prebakedDir = process.env.JACS_SMOKE_AGENT_DIR;
  if (prebakedDir) {
    if (!existsSync(prebakedDir)) return null;
    const prebaked = join(realpathSync(prebakedDir), 'jacs.config.json');
    if (existsSync(prebaked)) return prebaked;
    return null;
  }

  return createFreshJacsAgentConfig(workdir, 'remote-smoke-agent').configPath ?? null;
}

function resolveFreshJacsAgentConfig(
  workdir: string,
): { configPath?: string; skipReason?: string } {
  return createFreshJacsAgentConfig(workdir, 'local-smoke-agent');
}

/**
 * Create a fresh, signed-v2 JACS config with the current Node binding.
 *
 * The smoke tests intentionally do not shell out to an independently built
 * CLI: doing so can silently exercise an older JACS producer than the native
 * addon under test. Validating the emitted config here makes artifact skew a
 * visible skip reason instead of a confusing verification failure later.
 */
function createFreshJacsAgentConfig(
  workdir: string,
  name: string,
): { configPath?: string; skipReason?: string } {
  workdir = realpathSync(workdir);
  const password =
    process.env._HAISDK_SMOKE_PASSWORD ?? process.env.JACS_PRIVATE_KEY_PASSWORD ?? 'smoke-password';
  process.env.JACS_PRIVATE_KEY_PASSWORD = password;

  const configPath = join(workdir, 'jacs.config.json');
  try {
    const jacs = dynamicRequire('@hai.ai/jacs') as {
      createAgentSync: (
        name: string,
        password: string,
        algorithm?: string | null,
        dataDirectory?: string | null,
        keyDirectory?: string | null,
        configPath?: string | null,
        agentType?: string | null,
        description?: string | null,
        domain?: string | null,
        defaultStorage?: string | null,
      ) => string;
    };
    const result = JSON.parse(
      jacs.createAgentSync(
        name,
        password,
        'ring-Ed25519',
        join(workdir, 'data'),
        join(workdir, 'keys'),
        configPath,
        null,
        null,
        null,
        'fs',
      ),
    ) as { config_path?: string };
    const emittedConfigPath = result.config_path ?? configPath;
    if (!existsSync(emittedConfigPath)) {
      return {
        skipReason: `@hai.ai/jacs createAgentSync returned without writing ${emittedConfigPath}`,
      };
    }

    const config = JSON.parse(readFileSync(emittedConfigPath, 'utf8')) as {
      jacsSignature?: { signatureContentVersion?: string };
      jacs_default_storage?: string;
    };
    if (config.jacsSignature?.signatureContentVersion !== 'jacs-signature-v2') {
      return {
        skipReason:
          '@hai.ai/jacs emitted a config without jacs-signature-v2; rebuild the current binding',
      };
    }
    if (config.jacs_default_storage !== 'fs') {
      return {
        skipReason: `@hai.ai/jacs emitted unexpected storage ${String(config.jacs_default_storage)}`,
      };
    }

    return { configPath: realpathSync(emittedConfigPath) };
  } catch (error) {
    return {
      skipReason: `cannot bootstrap a fresh JACS agent with @hai.ai/jacs: ${String(error)}`,
    };
  }
}

describeWhenAvailable('haiinpm native FFI smoke test', () => {
  // Snapshot/restore JACS_DEFAULT_STORAGE around each test so per-test
  // overrides don't leak into sibling tests or the parent process. We use
  // explicit setup rather than vitest's `vi.stubEnv` because process.env
  // is what the Rust side reads — stubEnv operates only on import.meta.env.
  let savedDefaultStorage: string | undefined;
  beforeEach(() => {
    savedDefaultStorage = process.env.JACS_DEFAULT_STORAGE;
  });
  afterEach(() => {
    if (savedDefaultStorage === undefined) {
      delete process.env.JACS_DEFAULT_STORAGE;
    } else {
      process.env.JACS_DEFAULT_STORAGE = savedDefaultStorage;
    }
  });

  it(
    'saveMemory round-trips through the real native binding (remote backend)',
    async (ctx: TaskContext) => {
      if (!haiinpm) throw new Error(`haiinpm not loaded: ${String(loadError)}`);

      // Force remote routing for THIS test. Without this, the FFI's
      // `build_document_provider` falls through to `default_storage: "fs"`
      // (set in the generated config), routes to LocalJacsProvider, and never
      // makes the HTTP call this test was written to verify.
      process.env.JACS_DEFAULT_STORAGE = 'remote';

      type CapturedRequest = {
        method: string;
        url: string;
        headers: Record<string, string | string[] | undefined>;
        body: string;
      };
      const captured: CapturedRequest[] = [];

      const server = createServer((req: IncomingMessage, res: ServerResponse) => {
        let buffer = '';
        req.setEncoding('utf-8');
        req.on('data', (chunk) => {
          buffer += chunk;
        });
        req.on('end', () => {
          captured.push({
            method: req.method ?? '',
            url: req.url ?? '',
            headers: req.headers as Record<string, string | string[] | undefined>,
            body: buffer,
          });

          // The FFI's `save_memory(singleton: true)` first issues a GET
          // to /api/v1/records to check for an existing singleton. Reply
          // with an empty `items` list so the caller takes the
          // "no existing → create" branch and proceeds to POST.
          if (req.method === 'GET' && req.url?.startsWith('/api/v1/records')) {
            res.statusCode = 200;
            res.setHeader('Content-Type', 'application/json');
            res.end(JSON.stringify({ items: [], next_cursor: null }));
            return;
          }
          if (req.url === '/api/v1/records' && req.method === 'POST') {
            const payload = JSON.stringify({
              key: 'smoke:v1',
              id: 'smoke',
              version: 'v1',
              jacsType: 'memory',
              jacsVersionDate: '2026-01-01T00:00:00Z',
            });
            res.statusCode = 201;
            res.setHeader('Content-Type', 'application/json');
            res.end(payload);
          } else {
            res.statusCode = 404;
            res.end();
          }
        });
      });

      await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
      const address = server.address();
      if (!address || typeof address === 'string') throw new Error('no address');
      const baseUrl = `http://127.0.0.1:${address.port}`;

      const workdir = realpathSync(mkdtempSync(join(tmpdir(), 'haiai-smoke-remote-')));
      try {
        const configPath = resolveJacsAgentConfig(workdir);
        if (!configPath) {
          // eslint-disable-next-line no-console
          console.warn('remote smoke test skipped: cannot resolve JACS agent config');
          ctx.skip();
          return;
        }

        const ffiConfig = JSON.stringify({
          base_url: baseUrl,
          jacs_config_path: configPath,
          jacs_storage_backend: 'remote',
          client_type: 'node',
          timeout_secs: 5,
          max_retries: 0,
        });

        const HaiClientCtor = haiinpm.HaiClient;
        const client = new HaiClientCtor(ffiConfig) as {
          saveMemory: (content: string | null) => Promise<string>;
        };
        const key = await client.saveMemory('smoke-content');
        expect(key).toBe('smoke:v1');

        // The FFI does at least: 1 GET (find_document singleton check)
        // + 1 POST (sign+store). Assert the POST is what we expect; the
        // GET count can vary with future routing tweaks.
        const posts = captured.filter((r) => r.method === 'POST');
        expect(posts).toHaveLength(1);
        expect(posts[0].url).toBe('/api/v1/records');
        expect(String(posts[0].headers['content-type'])).toContain('text/markdown');
        expect(posts[0].body).toContain('smoke-content');
        expect(posts[0].body).toContain('-----BEGIN JACS SIGNATURE-----');
      } finally {
        server.close();
        rmSync(workdir, { recursive: true, force: true });
      }
    },
    30_000,
  );

  it(
    'saveMemory persists locally without HTTP traffic (fs backend)',
    async (ctx: TaskContext) => {
      if (!haiinpm) throw new Error(`haiinpm not loaded: ${String(loadError)}`);

      // Force fs routing. The pre-baked smoke agent already defaults to
      // fs, but pinning the env var here makes the test hermetic against
      // future bootstrap changes or a leaking parent-shell env var.
      process.env.JACS_DEFAULT_STORAGE = 'fs';

      const workdir = realpathSync(mkdtempSync(join(tmpdir(), 'haiai-smoke-local-')));
      try {
        const fresh = resolveFreshJacsAgentConfig(workdir);
        if (!fresh.configPath) {
          // eslint-disable-next-line no-console
          console.warn(`local smoke test skipped: ${fresh.skipReason}`);
          ctx.skip();
          return;
        }

        // No mock HTTP server: the local path must not make any network
        // calls, and binding the FFI to an unreachable URL surfaces that
        // invariant if the routing decision ever regresses.
        const ffiConfig = JSON.stringify({
          base_url: 'http://127.0.0.1:1', // unreachable on purpose
          jacs_config_path: fresh.configPath,
          jacs_storage_backend: 'fs',
          client_type: 'node',
          timeout_secs: 5,
          max_retries: 0,
        });

        const HaiClientCtor = haiinpm.HaiClient;
        const client = new HaiClientCtor(ffiConfig) as {
          saveMemory: (content: string | null) => Promise<string>;
          getRecordBytes: (key: string) => Promise<Buffer>;
        };
        const key = await client.saveMemory('local-smoke-content');

        expect(key).toMatch(LOCAL_KEY_PATTERN);

        // Round-trip: fetch the just-stored document by key. The FFI returns
        // the raw bytes of the signed text artifact, which must contain the
        // original plaintext we saved.
        const recordBytes: Buffer = await client.getRecordBytes(key);
        expect(Buffer.isBuffer(recordBytes)).toBe(true);
        expect(recordBytes.toString('utf8')).toContain('local-smoke-content');
      } finally {
        rmSync(workdir, { recursive: true, force: true });
      }
    },
    30_000,
  );
});

File: go/ffi_native_smoke_test.go
//go:build cgo_smoke

// Real-FFI smoke tests for haiigo (libhaiigo cdylib).
//
// Two tests, one per backend, both loading the real cgo binding and
// exercising `SaveMemory("...")` end-to-end:
//
//  1. `TestNativeSmokeSaveMemoryRoundTripsThroughLibhaiigo` (REMOTE) —
//     hosted production path. Sets `JACS_DEFAULT_STORAGE=remote` so the FFI
//     builds a `RemoteJacsProvider`, signs locally, POSTs to a real
//     `httptest.NewServer`, and reads the server-issued key from the
//     response. Verifies the mock saw a `POST /api/v1/records` with
//     signed markdown bytes (plaintext plus the JACS signature footer).
//
//  2. `TestNativeSmokeSaveMemoryLocalPath` (LOCAL) — dev default path
//     (`haiai init` writes `default_storage: "fs"`). Bootstraps a fresh
//     agent, sets `JACS_DEFAULT_STORAGE=fs`, signs locally, writes to disk,
//     and returns a client-side `{jacsId}:{jacsVersion}` key. Verifies the
//     doc round-trips via `GetRecordBytes(key)`.
//
// Together these cover the only two backends production and dev users
// actually exercise.
//
// Gated by the `cgo_smoke` build tag so it only runs when explicitly
// invoked:
//
//	go test -tags cgo_smoke -run NativeSmoke ./go/...
//
// Per PRD docs/haiai/JACS_DOCUMENT_STORE_FFI_PRD.md §5.5: real listener,
// no HTTP-level mock. The traffic is Rust `reqwest` inside libhaiigo,
// which only a real socket can intercept.

package haiai_test

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"runtime"
	"strings"
	"sync"
	"testing"

	"github.com/HumanAssisted/haiai-go/ffi"
)

// `LocalJacsProvider::store_signed_text` returns the key as
// `{jacsId}:{jacsVersion}` where both halves are JACS UUIDs. This regex
// matches that exact shape so the local-path test asserts on the key
// *structure* (not a specific value, which would change every run).
var localKeyPattern = regexp.MustCompile(
	`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}` +
		`:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

// TestNativeSmokeSaveMemoryRoundTripsThroughLibhaiigo exercises the REMOTE
// backend: `JACS_DEFAULT_STORAGE=remote` routes the FFI through
// `RemoteJacsProvider`, which signs locally and POSTs to the mock server.
func TestNativeSmokeSaveMemoryRoundTripsThroughLibhaiigo(t *testing.T) {
	// `t.Setenv` snapshots and restores the env var around the test, so
	// per-test routing overrides don't leak into sibling tests. Without
	// this, the FFI's `build_document_provider` falls through to
	// `default_storage: "fs"` (set by `haiai init`), routes to
	// LocalJacsProvider, and never makes the HTTP call this test was
	// written to verify.
	t.Setenv("JACS_DEFAULT_STORAGE", "remote")

	var (
		mu       sync.Mutex
		captured []map[string]any
	)

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Read the entire body so assertions are not flaky on partial reads.
		// `io.ReadAll` accumulates until EOF, which is required because
		// `r.Body.Read` is permitted to return fewer bytes than requested
		// per the `io.Reader` contract.
		body, err := io.ReadAll(r.Body)
		if err != nil {
			t.Logf("body read error: %v", err)
		}
		mu.Lock()
		captured = append(captured, map[string]any{
			"method":  r.Method,
			"path":    r.URL.Path,
			"content": r.Header.Get("Content-Type"),
			"body":    string(body),
		})
		mu.Unlock()

		// The FFI's `save_memory(singleton: true)` first issues a GET to
		// /api/v1/records to check for an existing singleton. Reply with
		// an empty `items` list so the caller takes the
		// "no existing → create" branch and proceeds to POST.
		if r.URL.Path == "/api/v1/records" && r.Method == http.MethodGet {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"items":[],"next_cursor":null}`))
			return
		}
		if r.URL.Path == "/api/v1/records" && r.Method == http.MethodPost {
			w.Header().Set("Content-Type", "application/json")
			w.WriteHeader(http.StatusCreated)
			_, _ = w.Write([]byte(`{"key":"smoke:v1","id":"smoke","version":"v1","jacsType":"memory","jacsVersionDate":"2026-01-01T00:00:00Z"}`))
			return
		}
		w.WriteHeader(http.StatusNotFound)
	}))
	defer server.Close()

	configPath := bootstrapJacsAgentOrSkip(t, "haiai-smoke-go-remote-")

	cfg := map[string]any{
		"base_url":             server.URL,
		"jacs_config_path":     configPath,
		"jacs_storage_backend": "remote",
		"client_type":          "go",
		"timeout_secs":         5,
		"max_retries":          0,
	}
	cfgJSON, err := json.Marshal(cfg)
	if err != nil {
		t.Fatalf("marshal config: %v", err)
	}

	client, err := ffi.NewClient(string(cfgJSON))
	if err != nil {
		t.Fatalf("ffi.NewClient: %v", err)
	}
	defer client.Close()

	key, err := client.SaveMemory("smoke-content")
	if err != nil {
		t.Fatalf("SaveMemory: %v", err)
	}
	if key != "smoke:v1" {
		t.Errorf("expected key=smoke:v1, got %q", key)
	}

	mu.Lock()
	defer mu.Unlock()
	// The FFI does at least: 1 GET (find_document singleton check) + 1
	// POST (sign+store). Assert the POST is what we expect; GET count can
	// vary with future routing tweaks.
	var posts []map[string]any
	for _, r := range captured {
		if r["method"] == "POST" {
			posts = append(posts, r)
		}
	}
	if len(posts) != 1 {
		t.Fatalf("expected exactly 1 POST to /api/v1/records, got %d (captured methods: %v)",
			len(posts), capturedMethods(captured))
	}
	req := posts[0]
	if req["path"] != "/api/v1/records" {
		t.Errorf("expected /api/v1/records, got %v", req["path"])
	}
	if !strings.Contains(req["content"].(string), "text/markdown") {
		t.Errorf("expected text/markdown content type, got %v", req["content"])
	}
	if !strings.Contains(req["body"].(string), "smoke-content") {
		t.Errorf("expected original plaintext in signed markdown body, got %v", req["body"])
	}
	if !strings.Contains(req["body"].(string), "-----BEGIN JACS SIGNATURE-----") {
		t.Errorf("expected JACS signature footer in signed markdown body, got %v", req["body"])
	}
}

func capturedMethods(reqs []map[string]any) []string {
	out := make([]string, 0, len(reqs))
	for _, r := range reqs {
		out = append(out, r["method"].(string)+" "+r["path"].(string))
	}
	return out
}

// TestNativeSmokeSaveMemoryLocalPath exercises the LOCAL (fs) backend: the
// FFI signs and writes to disk without any HTTP traffic. The pre-baked
// smoke agent already defaults to fs, but pinning the env var here makes
// the test hermetic against future bootstrap changes or a leaking
// parent-shell env var.
func TestNativeSmokeSaveMemoryLocalPath(t *testing.T) {
	t.Setenv("JACS_DEFAULT_STORAGE", "fs")

	configPath := bootstrapFreshJacsAgentOrSkip(t, "haiai-smoke-go-local-")

	// No mock HTTP server: the local path must not make any network
	// calls, and binding the FFI to an unreachable URL surfaces that
	// invariant if the routing decision ever regresses.
	cfg := map[string]any{
		"base_url":             "http://127.0.0.1:1", // unreachable on purpose
		"jacs_config_path":     configPath,
		"jacs_storage_backend": "fs",
		"client_type":          "go",
		"timeout_secs":         5,
		"max_retries":          0,
	}
	cfgJSON, err := json.Marshal(cfg)
	if err != nil {
		t.Fatalf("marshal config: %v", err)
	}

	client, err := ffi.NewClient(string(cfgJSON))
	if err != nil {
		t.Fatalf("ffi.NewClient: %v", err)
	}
	defer client.Close()

	key, err := client.SaveMemory("local-smoke-content")
	if err != nil {
		t.Fatalf("SaveMemory: %v", err)
	}
	if !localKeyPattern.MatchString(key) {
		t.Errorf("expected local key matching `{jacsId}:{jacsVersion}` UUID shape, got %q", key)
	}

	// Round-trip: fetch the just-stored document by key. The FFI returns
	// the raw bytes of the signed text artifact, which must contain the
	// original plaintext we saved.
	recordBytes, err := client.GetRecordBytes(key)
	if err != nil {
		t.Fatalf("GetRecordBytes: %v", err)
	}
	if !strings.Contains(string(recordBytes), "local-smoke-content") {
		t.Errorf("expected stored signed-text artifact to contain plaintext, got %q",
			string(recordBytes))
	}
}

func locateHaiaiCLI() (string, bool) {
	if explicit := os.Getenv("HAIAI_CLI"); explicit != "" && isExecutable(explicit) {
		return explicit, true
	}
	if _, file, _, ok := runtime.Caller(0); ok {
		if cli, ok := walkForHaiaiCLI(filepath.Dir(file)); ok {
			return cli, true
		}
	}
	if wd, err := os.Getwd(); err == nil {
		if cli, ok := walkForHaiaiCLI(wd); ok {
			return cli, true
		}
	}
	if cli, err := exec.LookPath("haiai"); err == nil {
		return cli, true
	}
	return "", false
}

func walkForHaiaiCLI(start string) (string, bool) {
	for dir := start; ; dir = filepath.Dir(dir) {
		candidate := filepath.Join(dir, "rust", "target", "release", haiaiBinaryName())
		if isExecutable(candidate) {
			return candidate, true
		}
		if _, err := os.Stat(filepath.Join(dir, ".git")); err == nil {
			return "", false
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return "", false
		}
	}
}

func haiaiBinaryName() string {
	if runtime.GOOS == "windows" {
		return "haiai.exe"
	}
	return "haiai"
}

func isExecutable(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir() && info.Mode()&0o111 != 0
}

func bootstrapFreshJacsAgentOrSkip(t *testing.T, tempPrefix string) string {
	t.Helper()

	cli, ok := locateHaiaiCLI()
	if !ok {
		t.Skipf("smoke test skipped: haiai CLI binary not found")
	}

	workdir, err := os.MkdirTemp("", tempPrefix)
	if err != nil {
		t.Fatalf("mkdtemp: %v", err)
	}
	if realWorkdir, err := filepath.EvalSymlinks(workdir); err == nil {
		workdir = realWorkdir
	}
	t.Cleanup(func() {
		_ = os.RemoveAll(workdir)
	})

	password := os.Getenv("_HAISDK_SMOKE_PASSWORD")
	if password == "" {
		password = os.Getenv("JACS_PRIVATE_KEY_PASSWORD")
	}
	if password == "" {
		password = "smoke-password"
		t.Setenv("JACS_PRIVATE_KEY_PASSWORD", password)
	}

	configPath := filepath.Join(workdir, "jacs.config.json")
	cmd := exec.Command(
		cli,
		"init",
		"--quiet",
		"--name",
		"local-smoke-agent",
		"--register",
		"false",
		"--data-dir",
		filepath.Join(workdir, "data"),
		"--key-dir",
		filepath.Join(workdir, "keys"),
		"--config-path",
		configPath,
	)
	cmd.Env = upsertEnv(os.Environ(), "JACS_PRIVATE_KEY_PASSWORD", password)
	output, err := cmd.CombinedOutput()
	if err != nil {
		t.Skipf("smoke test skipped: haiai init failed: %v; output=%s", err, string(output))
	}
	if _, err := os.Stat(configPath); err != nil {
		t.Skipf("smoke test skipped: haiai init did not write %s: %v", configPath, err)
	}
	return configPath
}

func upsertEnv(env []string, key string, value string) []string {
	prefix := key + "="
	for i, entry := range env {
		if strings.HasPrefix(entry, prefix) {
			env[i] = prefix + value
			return env
		}
	}
	return append(env, prefix+value)
}

// bootstrapJacsAgentOrSkip prepares the shared CI JACS agent for the remote
// smoke test or calls `t.Skip`. Local-path smoke tests use
// bootstrapFreshJacsAgentOrSkip so singleton lookups cannot observe state
// written by the remote signer.
//
// Returns the absolute path to jacs.config.json.
func bootstrapJacsAgentOrSkip(t *testing.T, tempPrefix string) string {
	t.Helper()
	_ = tempPrefix

	src := os.Getenv("JACS_SMOKE_AGENT_DIR")
	if src == "" {
		t.Skipf("smoke test skipped: JACS_SMOKE_AGENT_DIR not set — cannot bootstrap JACS agent")
	}
	srcConfig := filepath.Join(src, "jacs.config.json")
	if _, err := os.Stat(srcConfig); err != nil {
		t.Skipf("smoke test skipped: %s missing or unreadable: %v", srcConfig, err)
	}
	if realConfig, err := filepath.EvalSymlinks(srcConfig); err == nil {
		return realConfig
	}
	return srcConfig
}
